deltacat/compute/compactor_v2/utils/compaction_utils.py

```diff
@@ -0,0 +1,800 @@
+)
+from deltacat.compute.compactor_v2.model.hash_bucket_input import HashBucketInput
+
+from deltacat import logs
+from deltacat.compute.compactor_v2.model.merge_input import MergeInput
+from deltacat.compute.compactor_v2.model.merge_result import MergeResult
+from deltacat.compute.compactor_v2.model.hash_bucket_result import HashBucketResult
+from deltacat.compute.compactor.model.materialize_result import MaterializeResult
+from deltacat.compute.compactor_v2.utils.merge import (
+    generate_local_merge_input,
+)
+from deltacat.compute.compactor_v2.utils.task_options import (
+    hash_bucket_resource_options_provider,
+)
+from deltacat.compute.compactor.utils import round_completion_file as rcf
+from deltacat.compute.compactor import DeltaAnnotated
+from deltacat.compute.compactor_v2.utils.delta import contains_delete_deltas
+from deltacat.compute.compactor_v2.deletes.delete_strategy import (
+    DeleteStrategy,
+)
+from deltacat.compute.compactor_v2.deletes.delete_file_envelope import (
+    DeleteFileEnvelope,
+)
+from deltacat.utils.resources import (
+    get_current_process_peak_memory_usage_in_bytes,
+)
+from deltacat.compute.compactor_v2.deletes.utils import prepare_deletes
+
+from deltacat.storage import (
+    Delta,
+    DeltaType,
+    DeltaLocator,
+    Stream,
+    StreamLocator,
+    Partition,
+    Manifest,
+)
+from deltacat.compute.compactor.model.compact_partition_params import (
+    CompactPartitionParams,
+)
+from deltacat.utils.ray_utils.concurrency import (
+    invoke_parallel,
+    task_resource_options_provider,
+)
+from deltacat.compute.compactor_v2.steps import merge as mg
+from deltacat.compute.compactor_v2.steps import hash_bucket as hb
+from deltacat.compute.compactor_v2.utils import io
+
+from typing import List, Optional
+from collections import defaultdict
+from deltacat.compute.compactor.model.compaction_session_audit_info import (
+    CompactionSessionAuditInfo,
+)
+from deltacat.compute.compactor_v2.utils.task_options import (
+    merge_resource_options_provider,
+    local_merge_resource_options_provider,
+)
+
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+def _start_execute_compaction(
+    params: CompactPartitionParams,
+):
+    rcf_source_partition_locator = (
+        params.rebase_source_partition_locator or params.source_partition_locator
+    )
+
+    base_audit_url = rcf_source_partition_locator.path(
+        f"s3://{params.compaction_artifact_s3_bucket}/compaction-audit"
+    )
+    audit_url = f"{base_audit_url}.json"
+    logger.info(f"Compaction audit will be written to {audit_url}")
+    mutable_compaction_audit = (
+        CompactionSessionAuditInfo(deltacat.__version__, ray.__version__, audit_url)
+        .set_hash_bucket_count(params.hash_bucket_count)
+        .set_compactor_version(CompactorVersion.V2.value)
+    )
+
+    compaction_start = time.monotonic()
+
+    task_max_parallelism: int = params.task_max_parallelism
```

Comment:
The caller of this method already has access to the `params`, so it might not make sense to access and return `task_max_parallelism` here -- it should be possible to extract the max_parallelism later in compaction_session when you need it.